service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java

```diff
@@ -92,15 +97,32 @@ private boolean validateRequestPath(DefaultHttpSession acceptSession) {
         return requestPath.startsWith(acceptPath);
     }
 
+
+    /**
+     * Helper method performing loop detection
+     * @param acceptSession - session parameter
+     */
+    private void performLoopDetection(DefaultHttpSession acceptSession) {
+        List<String> viaHeaders = acceptSession.getReadHeaders(HEADER_VIA);
+        if (viaHeaders != null && viaHeaders.stream().anyMatch(h -> h.equals(viaHeader))) {
+                LOGGER.warn("Connection to " + getConnectURIs().iterator().next() +
+                        " failed due to loop detection [" + acceptSession + "->]");
+                acceptSession.setStatus(HttpStatus.SERVER_LOOP_DETECTED);
+                acceptSession.close(true);
+            }
+    }
+
     /*
      * Initializer for connect session. It adds the processed accept session headers
      * on the connect session
      */
     private static class ConnectSessionInitializer implements IoSessionInitializer<ConnectFuture> {
         private final DefaultHttpSession acceptSession;
+        private final String viaHeader;
 
-        ConnectSessionInitializer(DefaultHttpSession acceptSession) {
+        ConnectSessionInitializer(DefaultHttpSession acceptSession, String viaHeader) {
```

Comment:
It seems like ConnectSessionInitializer non-static class (that way you don't have to pass viaHeader parameter. similarly useForwarded parameter etc)